Documentation/network/lb-ipam.rst

```diff
@@ -0,0 +1,382 @@
+.. only:: not (epub or latex or html)
+
+    WARNING: You are looking at unreleased Cilium documentation.
+    Please use the official rendered version released here:
+    https://docs.cilium.io
+
+.. _lb_ipam:
+
+********************************************
+LoadBalancer IP Address Management (LB IPAM)
+********************************************
+
+LB IPAM is a feature that allows Cilium to assign IP addresses to Services of
+type ``LoadBalancer``. This functionally is usually left up to a cloud provider,
+however, when deploying in a private cloud environment, these facilities are not
+always available.
+
+LB IPAM work in conjunction with features like the :ref:`bgp_control_plane`. Where
+LB IPAM is responsible for allocation and assigning of IPs to Service objects and
+other features are responsible for load balancing and/or advertisement of these
+IPs. 
+
+LB IPAM is implicitly enabled when dependant features are enabled.
+
+Pools
+#####
+
+LB IPAM has the notion of IP Pools which the administrator can create to tell 
+Cilium which IP ranges can be used to allocate IPs from.
+
+A basic IP Pools with both an IPv4 and IPv6 range looks like.
+
+.. code-block:: yaml
+
+    apiVersion: "cilium.io/v2alpha1"
+    kind: CiliumLoadBalancerIPPool
+    metadata:
+      name: "blue-pool"
+    spec:
+      cidrs:
+      - cidr: "10.0.10.0/24"
+      - cidr: "2004::0/64"
+
+After adding the pool to the cluster, it appears like so.
+
+.. code-block:: shell-session
+
+    $ kubectl get ippools                           
+    NAME        DISABLED   CONFLICTING   IPS AVAILABLE   AGE
+    blue-pool   false      false         65788           2s
+
+.. note::
+    The amount of available IPs in the pool is less than the actual sum of all
+    usable IPs in the CIDRs because the allocation logic is limited to 65536 IPs
+    per CIDR. CIDRs containing more than 65536 IPs can be broken up into multiple
+    smaller CIDRs to achieve full utilization.
+
+Service Selectors
+-----------------
+
+IP Pools have an optional ``.spec.serviceSelector`` field which allows administrators
+to limit which services can get IPs from which pools using a `label selector <https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/>`__.
+The pool will allocate to any service if no service selector is specified.
+
+.. code-block:: yaml
+
+    apiVersion: "cilium.io/v2alpha1"
+    kind: CiliumLoadBalancerIPPool
+    metadata:
+      name: "blue-pool"
+    spec:
+      cidrs:
+      - cidr: "20.0.10.0/24"
+      serviceSelector:
```

Comment:
While we're at it, this should have a NamespaceSelector too. I would expect the semantics to be the same as with NetworkPolicy (predicates are ANDed).

Reply:
We do have a namespace selector, borrowed the concept from our security policies. Users can select on namespaces with the `io.kubernetes.service.namespace` label or in an expression if they need to allow multiple namespaces. This is explained a little bit further in the docs as well.